{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GRPH\n",
    "Given: A collection of DNA strings in FASTA format having total length at most 10 kbp.\n",
    "\n",
    "Return: The adjacency list corresponding to O3. You may return edges in any order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import networkx as nx\n",
    "\n",
    "def GRPH(input_file, k=3):\n",
    "    sequences = {record.id: str(record.seq) for record in SeqIO.parse(input_file, \"fasta\")}\n",
    "    graph = nx.DiGraph() # first we need to create a directed graph to represent the overlap relationships\n",
    "                         # each node in the graph corresponds to a sequence and edges represent suffix-prefix overlaps\n",
    "\n",
    "    # now we compare all pairs of sequences to find overlaps\n",
    "    for seq1_id, seq1_seq in sequences.items():    # for each sequence (source node)\n",
    "        for seq2_id, seq2_seq in sequences.items(): # it compares with every other sequence (target node)\n",
    "            if seq1_id != seq2_id and seq1_seq[-k:] == seq2_seq[:k]:  # we check if the suffix of seq1 matches the prefix of seq2 with length k and we make sure that seq1 != seq2 to prevent self-loops\n",
    "                graph.add_edge(seq1_id, seq2_id) # then we add a directed edge from seq1 to seq2\n",
    "    \n",
    "    adjacency = []    # the adjacency list contains all directed edges \n",
    "    for edge in graph.edges():\n",
    "        adjacency.append(f\"{edge[0]} {edge[1]}\")\n",
    "    \n",
    "    return adjacency\n",
    "\n",
    "input_file = \"rosalind_grph.txt\" \n",
    "k = 3  # we need to specify the overlap length\n",
    "\n",
    "adjacency = GRPH(input_file, k)\n",
    "for edge in adjacency:\n",
    "    print(edge)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TREE\n",
    "Given: A positive integer n (n≤1000) and an adjacency list corresponding to a graph on n nodes that contains no cycles.\n",
    "\n",
    "Return: The minimum number of edges that can be added to the graph to produce a tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "def TREE(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        input = file.readlines()\n",
    "    \n",
    "    nodes = int(input[0].strip()) # number of nodes given by the input\n",
    "    \n",
    "    graph = nx.Graph() # we initialize an undirected graph\n",
    "    graph.add_nodes_from(range(1, nodes + 1)) # then we add the nodes to the graph (from 1 to 'nodes')\n",
    "    \n",
    "    # we add the edges from the adjacency list (each line is an edge)\n",
    "    for line in input[1:]: \n",
    "        u, v = map(int, line.strip().split())\n",
    "        graph.add_edge(u, v)\n",
    "    \n",
    "    components = nx.number_connected_components(graph)  # we determine the n of connected components (a subgraph in which any two nodes are connected by a path)\n",
    "    min_edges = components - 1 # then we compute the minimum number of edges to connect all components into a tree\n",
    "    \n",
    "    return min_edges\n",
    "\n",
    "input_file = 'rosalind_tree.txt'\n",
    "result = TREE(input_file)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LONG\n",
    "Given: At most 50 DNA strings of approximately equal length, not exceeding 1 kbp, in FASTA format (which represent reads deriving from the same strand of a single linear chromosome).\n",
    "\n",
    "The dataset is guaranteed to satisfy the following condition: there exists a unique way to reconstruct the entire chromosome from these reads by gluing together pairs of reads that overlap by more than half their length.\n",
    "\n",
    "Return: A shortest superstring containing all the given strings (thus corresponding to a reconstructed chromosome)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def overlap(seq1, seq2):\n",
    "    combine_str = [] # merged strings\n",
    "    overlap_str = [] # overlapping strings\n",
    "\n",
    "    # we check if the suffix of seq1 matches the prefix of seq2\n",
    "    for i in range(len(seq1)):\n",
    "        if seq1[i:] == seq2[:len(seq1)-i]:  # overlap\n",
    "            overlap_str.append(seq1[i:])    # add the overlapping part\n",
    "            combine_str.append(seq1 + seq2[len(seq1)-i:]) # merge the strings\n",
    "            break\n",
    "\n",
    "    # we check if the suffix of s2 matches the prefix of s1\n",
    "    for i in range(len(seq2)):\n",
    "        if seq2[i:] == seq1[:len(seq2)-i]:  # overlap \n",
    "            overlap_str.append(seq2[i:])\n",
    "            combine_str.append(seq2 + seq1[len(seq2)-i:])\n",
    "            break\n",
    "\n",
    "    if not overlap_str:\n",
    "        return \"\", \"\"  # no overlap found, we return empty values\n",
    "\n",
    "    # then we rturn the shortest merged string and the longest overlap string\n",
    "    min_merge = min(combine_str, key=len)  \n",
    "    max_overlap = max(overlap_str, key=len) \n",
    "    return min_merge, max_overlap\n",
    "\n",
    "def superstring(strings):\n",
    "    list = strings\n",
    "\n",
    "    while len(list) > 1:\n",
    "        most_overlap = \"\"\n",
    "        most_pair = []\n",
    "        most_length = 0\n",
    "\n",
    "        # we iterate over all pairs of strings to find the pair with the maximum overlap\n",
    "        for i in range(len(list)-1):\n",
    "            for j in range(i+1, len(list)):\n",
    "                min_merge, max_overlap = overlap(list[i], list[j])\n",
    "                if len(max_overlap) > most_length:\n",
    "                    most_overlap = min_merge\n",
    "                    most_pair = [list[i], list[j]]\n",
    "                    most_length = len(max_overlap)\n",
    "\n",
    "        # we remove the overlapping pair \n",
    "        list.remove(most_pair[0])\n",
    "        list.remove(most_pair[1])\n",
    "        list.append(most_overlap) # and add the merged string\n",
    "\n",
    "    # the final remaining string is the shortest superstring\n",
    "    return list[0]\n",
    "\n",
    "def LONG(input_file):\n",
    "    seq_str = [str(record.seq) for record in SeqIO.parse(input_file, \"fasta\")]\n",
    "    return superstring(seq_str)\n",
    "\n",
    "input_file = \"rosalind_long.txt\"\n",
    "result = LONG(input_file)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DEG\n",
    "In an undirected graph, the degree d(u) of a vertex u is the number of neighbors u has, or equivalently, the number of edges incident upon it.\n",
    "\n",
    "Given: A simple graph with n≤10^3 vertices in the edge list format.\n",
    "\n",
    "Return: An array D[1..n] where D[i] is the degree of vertex i."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DEG(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        vertices, edges = map(int, file.readline().split()) # first we read the n of vertices and edges from the graph in the file\n",
    "        \n",
    "        # then we initialize an array (with zeros) to store the degree of each vertex\n",
    "        deg = [0] * vertices  \n",
    "\n",
    "        # we have to process each edge and update the degree for both vertices\n",
    "        for line in file:\n",
    "            u, v = map(int, line.split())\n",
    "            deg[u - 1] += 1 \n",
    "            deg[v - 1] += 1\n",
    "\n",
    "    return deg # this returns the degrees as a string so in the print() I added *\n",
    "\n",
    "input_file = 'rosalind_deg.txt'  \n",
    "result = DEG(input_file)\n",
    "print(*result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DDEG\n",
    "Given: A simple graph with n≤10^3 vertices in the edge list format.\n",
    "\n",
    "Return: An array D[1..n] where D[i] is the sum of the degrees of i's neighbors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DDEG(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        vertices, edges = map(int, file.readline().split()) # first we read the n of vertices and edges from the graph in the file\n",
    "        \n",
    "        # then we initialize an array to store the degree of each vertex\n",
    "        deg = [0] * vertices  \n",
    "        adjacency = [[] for i in range(vertices)]\n",
    "        \n",
    "        # then we need to process each edge and update the degree and adjacency list for both vertices\n",
    "        for line in file:\n",
    "            u, v = map(int, line.split())\n",
    "            u -= 1\n",
    "            v -= 1\n",
    "            \n",
    "            # here we update degree for both vertices\n",
    "            deg[u] += 1\n",
    "            deg[v] += 1\n",
    "            \n",
    "            # we add each vertex to the other's adjacency list\n",
    "            adjacency[u].append(v)\n",
    "            adjacency[v].append(u)\n",
    "        \n",
    "        # we need to compute the sum of the degrees of the neighbors for each vertex\n",
    "        result = []\n",
    "        for i in range(vertices):\n",
    "            neighbor_degree = sum(deg[neighbor] for neighbor in adjacency[i])\n",
    "            result.append(neighbor_degree)\n",
    "\n",
    "    return result\n",
    "\n",
    "input_file = 'rosalind_ddeg.txt'  \n",
    "result = DDEG(input_file)\n",
    "print(*result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CC\n",
    "Given: A simple graph with n≤10^3 vertices in the edge list format.\n",
    "\n",
    "Return: The number of connected components in the graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filereader(input_file):\n",
    "    with open(input_file, 'r') as f:\n",
    "        vertices, n_edges = map(int, f.readline().split())  # we read the n of vertices and edges\n",
    "        edges = [tuple(map(int, line.split())) for line in f]  # we read all edges\n",
    "    return vertices, n_edges, edges\n",
    "\n",
    "def search(node, graph, visit):\n",
    "    visit[node] = True              # we mark the the current node as visited\n",
    "    for neighbor in graph[node]:    # then we explore the neighbors\n",
    "        if not visit[neighbor]:     # if a neighbor hasn't been visited we recursively perform the depth-first search\n",
    "            search(neighbor, graph, visit)\n",
    "\n",
    "def CC(vertices, edges):\n",
    "    graph = {i: [] for i in range(1, vertices + 1)} # we create the graph as an adjacency list\n",
    "    for u, v in edges:\n",
    "        graph[u].append(v)\n",
    "        graph[v].append(u)  \n",
    "\n",
    "    visit = [False] * (vertices + 1)  # then we create an array to keep track of visited nodes\n",
    "    counter = 0  # count of connected components\n",
    "\n",
    "    # then we run the depth-first search for each unvisited node to identify connected components\n",
    "    for node in range(1, vertices + 1):\n",
    "        if not visit[node]: # if the current node hasn't been visited we perform the depth-first search from this node\n",
    "            search(node, graph, visit)\n",
    "            counter += 1  # and we increment the counter for a new connected component\n",
    "\n",
    "    return counter\n",
    "\n",
    "input_file = 'rosalind_cc.txt'  \n",
    "vertices, n_edges, edges = filereader(input_file)\n",
    "result = CC(vertices, edges)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BFS\n",
    "Given: A simple directed graph with n≤10^3 vertices in the edge list format.\n",
    "\n",
    "Return: An array D[1..n] where D[i] is the length of a shortest path from the vertex 1 to the vertex i (D[1]=0). \n",
    "If i is not reachable from 1 set D[i] to −1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "def BFS(input_file):\n",
    "    graph = nx.DiGraph()         # we first create an empty directed graph\n",
    "\n",
    "    with open(input_file, 'r') as file:\n",
    "        vertices, edges = map(int, file.readline().split())\n",
    "        \n",
    "        # now we add the edges to the graph\n",
    "        for line in file:\n",
    "            u, v = map(int, line.split())\n",
    "            graph.add_edge(u, v)\n",
    "\n",
    "    # we need to initialize the distances (-1 means unreachable)\n",
    "    dist = [-1] * vertices \n",
    "    dist[0] = 0  # because the distance from vertex 1 to itself is 0\n",
    "\n",
    "    visit = [1]  # we start the breadth-first search from vertex 1 (1-based index, so we add 1)\n",
    "    \n",
    "    while visit:\n",
    "        node = visit.pop(0)  # we pop the first element from 'visit'\n",
    "        \n",
    "        # we need to check all neighbors of the current node\n",
    "        for neighbor in graph.neighbors(node):\n",
    "            if dist[neighbor - 1] == -1:  # if the neighbor has not been visited\n",
    "                dist[neighbor - 1] = dist[node - 1] + 1\n",
    "                visit.append(neighbor)  # we add the neighbor to 'visit'\n",
    "    \n",
    "    return dist\n",
    "\n",
    "input_file = 'rosalind_bfs.txt' \n",
    "result = BFS(input_file)\n",
    "print(*result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DIJ\n",
    "Given: A simple directed graph with positive edge weights from 1 to 10^3 and n≤10^3 vertices in the edge list format.\n",
    "\n",
    "Return: An array D[1..n] where D[i] is the length of a shortest path from the vertex 1 to the vertex i (D[1]=0). If i is not reachable from 1 set D[i] to −1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DIJ(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "    vertices, n_edges = map(int, lines[0].split()) \n",
    "    edges = [tuple(map(int, line.split())) for line in lines[1:]]\n",
    "\n",
    "    # we initialize the graph as an adjacency list\n",
    "    graph = {node: [] for node in range(1, vertices + 1)}\n",
    "    for start, end, weight in edges:\n",
    "        graph[start].append((end, weight))  # then we add the directed edge from start to end with weight\n",
    "\n",
    "    dist = {node: float('inf') for node in range(1, vertices + 1)} # we initialize  the distances by putting infinity for all vertices except the source (vertex 1)\n",
    "    dist[1] = 0  # cause the distance from vertex 1 to itself is 0\n",
    "    list = [(0, 1)]  # a list to store tuples of (vertix, distance) and we start with vertex 1, distance 0\n",
    "    \n",
    "    while list:\n",
    "        current_dist, current_node = list.pop(0) # we pop the node with the smallest known distance\n",
    "        \n",
    "        if current_dist > dist[current_node]:  # we move on if the current distance is not the smallest known distance for the node\n",
    "            continue\n",
    "\n",
    "        # then we update the distances for all neighbors of the current node\n",
    "        for neighbor, weight in graph[current_node]:\n",
    "            new_dist = current_dist + weight\n",
    "            if new_dist < dist[neighbor]:\n",
    "                dist[neighbor] = new_dist\n",
    "                list.append((new_dist, neighbor))\n",
    "                list = sorted(list)  \n",
    "                \n",
    "    # in the result we replace unreachable distances (inf) with -1\n",
    "    result = [dist[node] if dist[node] != float('inf') else -1 for node in range(1, vertices + 1)]   \n",
    "    return result\n",
    "\n",
    "input_file = 'rosalind_dij.txt'  \n",
    "result = DIJ(input_file)\n",
    "print(\" \".join(map(str, result)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DAG\n",
    "Given: A positive integer k≤20 and k simple directed graphs in the edge list format with at most 10^3 vertices and 3⋅10^3 edges each.\n",
    "\n",
    "Return: For each graph, output \"1\" if the graph is acyclic and \"-1\" otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "def DAG(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "    n_graphs = int(lines[0].strip()) # we read the n of graphs\n",
    "\n",
    "    result = []\n",
    "    idx = 1\n",
    "\n",
    "    for i in range(n_graphs):                                  # we need to process each graph in the file\n",
    "        while idx < len(lines) and lines[idx].strip() == \"\":   # we skip any empty lines to reach the next graph's data\n",
    "            idx += 1\n",
    "\n",
    "        vertices, edges = map(int, lines[idx].strip().split()) # we read the n of vertices and edges\n",
    "        idx += 1\n",
    "\n",
    "        list_edges = []\n",
    "        for i in range(edges):\n",
    "            u, v = map(int, lines[idx].strip().split())\n",
    "            list_edges.append((u, v)) # we append each edge (parsed) as a tuple to the list \n",
    "            idx += 1\n",
    "\n",
    "        graph = nx.DiGraph()\n",
    "        graph.add_edges_from(list_edges) # we add our list of edges to the graph\n",
    "\n",
    "        # and we check if the graph is acyclic or not\n",
    "        if nx.is_directed_acyclic_graph(graph):\n",
    "            result.append(\"1\")\n",
    "        else:\n",
    "            result.append(\"-1\")\n",
    "\n",
    "    return \" \".join(result)\n",
    "\n",
    "input_file = \"rosalind_dag.txt\"\n",
    "result = DAG(input_file)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BIP\n",
    "Given: A positive integer k≤20 and k simple graphs in the edge list format with at most 10^3vertices each.\n",
    "\n",
    "Return: For each graph, output \"1\" if it is bipartite and \"-1\" otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "def BIP(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "    n_graphs = int(lines[0].strip())\n",
    "\n",
    "    result = []\n",
    "    idx = 1\n",
    "\n",
    "    for i in range(n_graphs):\n",
    "        while idx < len(lines) and lines[idx].strip() == \"\":\n",
    "            idx += 1\n",
    "\n",
    "        vertices, edges = map(int, lines[idx].strip().split()) # we read the n of vertices and edges\n",
    "        idx += 1\n",
    "        list_edges = []\n",
    "        for i in range(edges):\n",
    "            u, v = map(int, lines[idx].strip().split())\n",
    "            list_edges.append((u, v)) # we add the edges to the list as a tuple\n",
    "            idx += 1\n",
    "\n",
    "        graph = nx.Graph()\n",
    "        graph.add_edges_from(list_edges) # we add our list of edges to a graph\n",
    "\n",
    "        # and we check if the graph is bipartite or not\n",
    "        if nx.is_bipartite(graph):\n",
    "            result.append(\"1\")\n",
    "        else:\n",
    "            result.append(\"-1\")\n",
    "\n",
    "    return \" \".join(result)\n",
    "\n",
    "input_file = \"rosalind_bip.txt\"\n",
    "result = BIP(input_file)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BF\n",
    "Given: A simple directed graph with integer edge weights from −10^3 to 10^3 and n≤10^3 vertices in the edge list format.\n",
    "\n",
    "Return: An array D[1..n] where D[i] is the length of a shortest path from the vertex 1 to the vertex i (D[1]=0). If i is not reachable\n",
    "from 1 set D[i] to x."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def BF(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "    vertices, n_edges = map(int, lines[0].strip().split())\n",
    "\n",
    "    edges = []  # we initialize an empty list to store the edges (u, v, weight)\n",
    "    for line in lines[1:]:\n",
    "        u, v, weight = map(int, line.strip().split())\n",
    "        edges.append((u, v, weight))\n",
    "\n",
    "    # we initialize a list to store the shortest distances and set all distances to infinity at first\n",
    "    dist = [float('inf')] * (vertices + 1)\n",
    "    dist[1] = 0  # cause the distance to the source vertex is 0\n",
    "\n",
    "    # this is the Bellman-Ford algorithm\n",
    "    for i in range(vertices - 1):\n",
    "        for u, v, weight in edges:\n",
    "            if dist[u] != float('inf') and dist[u] + weight < dist[v]: # we check if the distance to vertex u is not infinity (it has been visited) and if a shorter path is found\n",
    "                dist[v] = dist[u] + weight                             # we 'relax' the edge by updating the distance to vertex v\n",
    "\n",
    "    # now we need to format the output and replace the unreachable vertices with 'x'\n",
    "    result = []\n",
    "    for i in range(1, vertices + 1):\n",
    "        if dist[i] == float('inf'): \n",
    "            result.append('x')\n",
    "        else:\n",
    "            result.append(str(dist[i]))\n",
    "\n",
    "    return \" \".join(result)\n",
    "\n",
    "input_file = \"rosalind_bf.txt\"\n",
    "result = BF(input_file)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CTE\n",
    "Given: A positive integer k≤20 and k simple directed graphs with positive integer edge weights and at most 10^3 vertices in the edge\n",
    "list format.\n",
    "\n",
    "Return: For each graph, output the length of a shortest cycle going through the first specified edge if there is a cycle and \"-1\"\n",
    "otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CTE(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    n_graphs = int(lines[0].strip())\n",
    "    result = []\n",
    "    idx = 1\n",
    "\n",
    "    for i in range(n_graphs):                                # we loop through each graph\n",
    "        while idx < len(lines) and lines[idx].strip() == \"\": # we skip empty lines between graphs\n",
    "            idx += 1\n",
    "\n",
    "        vertices, n_edges = map(int, lines[idx].strip().split()) # we read the n of vertices and edges for the current graph\n",
    "        idx += 1\n",
    "\n",
    "        edges = []\n",
    "        graph = [[] for i in range(vertices + 1)]\n",
    "        for i in range(n_edges):\n",
    "            u, v, weight = map(int, lines[idx].strip().split())\n",
    "            edges.append((u, v, weight))\n",
    "            graph[u].append((v, weight))\n",
    "            idx += 1\n",
    "\n",
    "        # we get the first edge to check if it participates in a cycle\n",
    "        first_edge = edges[0]\n",
    "        u1, v1, weight1 = first_edge\n",
    "\n",
    "        # we temporarily remove the first edge from the graph to check if it creates a cycle later\n",
    "        graph[u1] = [(v, weight) for v, weight in graph[u1] if v != v1 or weight != weight1]\n",
    "\n",
    "        dist = [float('inf')] * (vertices + 1)\n",
    "        dist[v1] = 0\n",
    "        visit = [v1]\n",
    "\n",
    "        # we need to use a sort of like BFS approach to find the shortest path\n",
    "        while visit:\n",
    "            u = visit.pop(0)\n",
    "            for neighbor, weight in graph[u]:\n",
    "                if dist[u] + weight < dist[neighbor]:\n",
    "                    dist[neighbor] = dist[u] + weight\n",
    "                    visit.append(neighbor)\n",
    "\n",
    "        # then we check if a cycle exists\n",
    "        if dist[u1] == float('inf'):\n",
    "            result.append(\"-1\")\n",
    "        else:\n",
    "            result.append(str(dist[u1] + weight1))\n",
    "\n",
    "    return \" \".join(result)\n",
    "\n",
    "input_file = \"rosalind_cte.txt\"\n",
    "result = CTE(input_file)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NWC\n",
    "Given: A positive integer k≤20 and k simple directed graphs with integer edge weights from −10^3 to 10^3 and n≤10^3 vertices in the edge list format.\n",
    "\n",
    "Return: For each graph, output \"1\" if it contains a negative weight cycle and \"-1\" otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import networkx as nx\n",
    "\n",
    "def neg_cycle(n, edges):\n",
    "    graph = nx.DiGraph()\n",
    "    graph.add_node(-1)                   # first we add a super source node (-1)\n",
    "    for i in range(1, n + 1):\n",
    "        graph.add_node(i)\n",
    "        graph.add_edge(-1, i, weight=0)  # then we connect the super source to all nodes with 0 weight\n",
    "    graph.add_weighted_edges_from(edges) # lastly we add all edges\n",
    "\n",
    "    try:\n",
    "        nx.single_source_bellman_ford_path_length(graph, -1)  # we run Bellman-Ford from super source node\n",
    "    except nx.NetworkXUnbounded:\n",
    "        return \"1\"  # if a negative cycle exists, it will raise a NetworkXUnbounded exception\n",
    "    return \"-1\"  # no negative cycle\n",
    "\n",
    "def NWC(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        lines = file.read().splitlines()\n",
    "\n",
    "    n_graphs = int(lines[0])  \n",
    "    result = []\n",
    "    i = 1                   # we start from the next line after the number of graphs\n",
    "    while i < len(lines):\n",
    "        vertices, n_edges = map(int, lines[i].split())\n",
    "        i += 1\n",
    "        edges = [tuple(map(int, lines[i + j].split())) for j in range(n_edges)]\n",
    "        i += n_edges\n",
    "        result.append(neg_cycle(vertices, edges)) # for the current graph, we call the neg_cycle function to check for negative cycles\n",
    "    \n",
    "    print(\" \".join(result))  \n",
    "\n",
    "input_file = 'rosalind_nwc.txt'\n",
    "NWC(input_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SDAG\n",
    "\n",
    "Given: A weighted DAG with integer edge weights from −103 to 103 and n≤105 vertices in the edge list format.\n",
    "\n",
    "Return: An array D[1..n] where D[i] is the length of a shortest path from the vertex 1 to the vertex i (D[1]=0). If i is not reachable from 1 set D[i] to x."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def topological(vertices, adjacency):\n",
    "    visit = [False] * (vertices + 1)\n",
    "    order = []\n",
    "\n",
    "    def depthsearch(v):\n",
    "        visit[v] = True\n",
    "        for neighbor, weight in adjacency.get(v, []):  # we use get to handle missing keys\n",
    "            if not visit[neighbor]:\n",
    "                depthsearch(neighbor)\n",
    "        order.append(v)\n",
    "\n",
    "    for i in range(1, vertices + 1):\n",
    "        if not visit[i]:\n",
    "            depthsearch(i)\n",
    "\n",
    "    return order[::-1]  # we need to reverse it to get topological order\n",
    "\n",
    "def shortest(vertices, adjacency):\n",
    "    dist = ['x'] * (vertices + 1) # first we initialize a distance array with 'x' for unreachable vertices\n",
    "    dist[1] = 0                   # cause the distance to the source vertex is 0\n",
    "\n",
    "    topo_order = topological(vertices, adjacency) # we get the topological order\n",
    "\n",
    "    # then we need to  'relax' the edges in the topological order\n",
    "    for u in topo_order:\n",
    "        if dist[u] == 'x':  # we want to skip unreachable nodes\n",
    "            continue\n",
    "        for v, weight in adjacency.get(u, []):      # we need to use get to handle missing keys\n",
    "            if dist[v] == 'x' or dist[v] > dist[u] + weight:\n",
    "                dist[v] = dist[u] + weight\n",
    "\n",
    "    return dist[1:]  # lastly, we return the distances for vertices 1 to n\n",
    "\n",
    "def SDAG(input_file):\n",
    "    with open(input_file, 'r') as file:\n",
    "        vertices, n_edges = map(int, file.readline().split())\n",
    "        \n",
    "        adjacency = {} # we initialize an empty dictionary to create an adjacency list\n",
    "        for i in range(n_edges):\n",
    "            u, v, w = map(int, file.readline().split())\n",
    "            if u not in adjacency:\n",
    "                adjacency[u] = []        # we need to initialize an empty list if vertex u is not in the dictionary\n",
    "            adjacency[u].append((v, w))  # and we append the edge to the adjacency list of u\n",
    "    \n",
    "    result = shortest(vertices, adjacency)\n",
    "    print(\" \".join(map(str, result)))\n",
    "\n",
    "\n",
    "SDAG('rosalind_sdag.txt')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
